### PR TITLE
feat: add dotnet

### LINKED ISSUE #53: Add dotnet
As per discussion in https://github.com/renovatebot/renovate/issues/6610

### PR DESCRIPTION
**Pull request by fgreinacher**
Fixes #53

### CODE REVIEW

**Comment by fgreinacher** [comment]
@viceice Is that enough? I just tried to mirror what you did for other tools. Should we also add a dedicated test repo or are the inline files here enough?

**Comment by viceice** [comment]
I'll check it tomorrow morining 🙈

File: src/dotnet/build/dotnet.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+set -e
+
+check_semver $DOTNET_VERSION
+
+if [[ ! "${MAJOR}" || ! "${MINOR}" || ! "${PATCH}" ]]; then
+  echo Invalid version: ${DOTNET_VERSION}
+  exit 1
+fi
+
+DOTNET_INSTALL_DIR=/usr/local/dotnet-${DOTNET_VERSION}
```
Comment (viceice) [inline src/dotnet/build/dotnet.sh:12]:
```suggestion
DOTNET_INSTALL_DIR=/usr/local/dotnet/${DOTNET_VERSION}
```

File: src/dotnet/build/dotnet.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+if [[ -d "${DOTNET_INSTALL_DIR}" ]]; then
+  echo "Skipping, already installed"
+  exit 0
+fi
+
+apt_install libc6 libgcc1 libgssapi-krb5-2 libicu60 libssl1.1 libstdc++6 zlib1g
```
Comment (viceice) [inline src/dotnet/build/dotnet.sh:19]:
```suggestion

mkdir -p /usr/local/dotnet
apt_install libc6 libgcc1 libgssapi-krb5-2 libicu60 libssl1.1 libstdc++6 zlib1g
```

File: src/dotnet/build/dotnet.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+curl -sSL -o /usr/local/dotnet-install.sh https://dot.net/v1/dotnet-install.sh
+bash /usr/local/dotnet-install.sh --version $DOTNET_VERSION --install-dir ${DOTNET_INSTALL_DIR}
```
Comment (viceice) [inline src/dotnet/build/dotnet.sh:22]:
```suggestion
curl -sSL https://dot.net/v1/dotnet-install.sh | bash - --version $DOTNET_VERSION --install-dir ${DOTNET_INSTALL_DIR}
```
do not store script

**Review by viceice — COMMENTED** [review]
just some small changes

File: src/dotnet/build/dotnet.sh [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+if [[ -d "${DOTNET_INSTALL_DIR}" ]]; then
+  echo "Skipping, already installed"
+  exit 0
+fi
+
+
+mkdir -p /usr/local/dotnet
+apt_install libc6 libgcc1 libgssapi-krb5-2 libicu60 libssl1.1 libstdc++6 zlib1g
+
+curl -sSL https://dot.net/v1/dotnet-install.sh | bash - --version $DOTNET_VERSION --install-dir ${DOTNET_INSTALL_DIR}
```
Comment (viceice) [inline src/dotnet/build/dotnet.sh:23]:
```suggestion
curl -sSfL https://dot.net/v1/dotnet-install.sh | sh -s - --version $DOTNET_VERSION --install-dir ${DOTNET_INSTALL_DIR}
```

File: src/dotnet/build/dotnet.sh [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+if [[ -d "${DOTNET_INSTALL_DIR}" ]]; then
+  echo "Skipping, already installed"
+  exit 0
+fi
+
+
+mkdir -p /usr/local/dotnet
+apt_install libc6 libgcc1 libgssapi-krb5-2 libicu60 libssl1.1 libstdc++6 zlib1g
+
+curl -sfL https://dot.net/v1/dotnet-install.sh | sh -s - --version $DOTNET_VERSION --install-dir ${DOTNET_INSTALL_DIR}
```
Comment (viceice) [inline src/dotnet/build/dotnet.sh:23]:
```suggestion
curl -sfL https://dot.net/v1/dotnet-install.sh | bash -s - --version $DOTNET_VERSION --install-dir ${DOTNET_INSTALL_DIR}
```

**Comment by viceice** [comment]
Are we missing some deps ?
image

**Comment by fgreinacher** [comment]
> Are we missing some deps ?
> image

Not sure, https://github.com/renovatebot/docker-buildpack/pull/54#issuecomment-709247779 states exactly the packages we install.

**Comment by viceice** [comment]
ok, so this is good now. but we can only install as root for now, so no dynamic install. 

if we want to install on the fly, we need to prepare another script to install as user.

**Comment by rarkins** [comment]
I thought when I originally looked into this that there'd be a part static (installed as root) / part dynamic (runtime) install. Is that not the case and we'll need to prebuild every nuget version into its own image?

**Comment by viceice** [comment]
It is possible to install multiple different versions at runtime. But our install-tool script checks for root, so we need a second one to install as user to eg `~/.dotnet/` folder.

**Comment by rarkins** [comment]
Don't think we need multiple? But it's usually helpful if we can install dynamically instead of needing to pre build every version into its own image

**Comment by viceice** [comment]
I think we can maybe need multiple versions at same time if a project is targeting multiple dotnet versions. I'll test this.

**Comment by fgreinacher** [comment]
> ok, so this is good now. but we can only install as root for now, so no dynamic install.
> 
> if we want to install on the fly, we need to prepare another script to install as user.

Is this a general limitation of the current approach or can I do something to help?

**Comment by viceice** [comment]
image

🙈

**Comment by fgreinacher** [comment]
> I think we can maybe need multiple versions at same time if a project is targeting multiple dotnet versions. I'll test this.

Usually newer SDKs can handle older target frameworks ("dotnet versions"). Actually running the applications (or tests) might be a problem, but this does not matter for our use case here.

**Comment by viceice** [comment]
yes, i've tested it the last commit works on restore with dotnet 3.1.x, even if we target net5.0

**Comment by viceice** [comment]
We will add dynamic install later, if we need it. for now it seems to be enough to have a single version to restore packages.

**Pull request merged**